File: algorithms/2_medium/L/01218_longest_arithmetic_subsequence_of_given_difference/table.c
#define MAX_ALOWED_NUM 10000
#define MIN_ALOWED_NUM -10000
#define MIN(a, b) (((a) < (b)) ? (a) : (b))
#define MAX(a, b) (((a) > (b)) ? (a) : (b))

int longestSubsequence(int* arr, const int len, const int difference) {
    assert(len > 0 && arr != NULL);
    if (len == 1) { return 1; }

    int max = INT_MIN;
    int min = INT_MAX;

    for (int i = 0; i < len; i += 1) {
        max = MAX(max, arr[i]);
        min = MIN(min, arr[i]);
    }
    assert(min >= MIN_ALOWED_NUM && max <= MAX_ALOWED_NUM);

    const size_t table_len = (size_t)(max - min + 1);
    int* table = calloc(table_len, sizeof(*table));
    assert(table != NULL);
    int res = 0;

    for (int i = 0; i < len; i += 1) {
        const int a = arr[i];
        int curr = 1;
        if (a - difference >= min && a - difference <= max) {
            const size_t idx = (size_t)(a - difference - min);
            curr = 1 + table[idx];
        }

        table[a - min] = curr;
        res = MAX(res, curr);
    }

/* cleanup: */
    free(table);
    return res;
}

File: algorithms/2_medium/L/01218_longest_arithmetic_subsequence_of_given_difference/table.rs
const MAX_ALOWED_NUM: i32 = 10_000;
const MIN_ALOWED_NUM: i32 = -10_000;

impl Solution {
    pub fn longest_subsequence(arr: Vec<i32>, difference: i32) -> i32 {
        assert!(arr.len() > 0);
        if arr.len() == 1 { return 1; }

        let mut max = i32::MIN;
        let mut min = i32::MAX;

        for &a in &arr {
            max = max.max(a);
            min = min.min(a);
        }
        assert!(min >= MIN_ALOWED_NUM && max <= MAX_ALOWED_NUM);

        let table_len = (max - min + 1) as usize;
        let mut table = vec![0_i32; table_len];
        let mut res = 0_i32;

        for a in arr {
            let mut curr = 1_i32;
            if a - difference >= min && a - difference <= max {
                let idx = (a - difference - min) as usize;
                curr = 1 + table[idx];
            }

            table[(a - min) as usize] = curr;
            res = res.max(curr);
        }

        res
    }
}
